src/lua/uri.lua

```diff
@@ -60,6 +60,41 @@ local uri_set_stash = buffer.ffi_stash_new('struct uri_set')
 local uri_set_stash_take = uri_set_stash.take
 local uri_set_stash_put = uri_set_stash.put
 
+-- RFC 3986: Uniform Resource Identifier (URI): Generic Syntax
+-- 2.2. Reserved Characters
+-- https://datatracker.ietf.org/doc/html/rfc3986.html#section-2.2
+local UNRESERVED_RFC3986 = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-._~"
+
+local RFC3986 = {
+    unreserved = UNRESERVED_RFC3986,
+    plus = false,
+}
+
+local FORM_DATA = {
+    unreserved = UNRESERVED_RFC3986,
+    plus = true,
+}
+
+local PATH_SEGMENT = {
+    unreserved = UNRESERVED_RFC3986 .. ":@-._~!$&'()*+,;=",
+    plus = false,
+}
+
+local PATH = {
+    unreserved = UNRESERVED_RFC3986 .. ":@-._~!$&'()*+,;=/",
+    plus = false,
+}
```

Comment:
AFAIR, we decided to add `PATH` aside of normative `PATH_SEGMENT` to allow to encode a path (like `/foo/bar`) directly, not per component.

To illustrate the idea, the ideal code would do the following:

```lua
local uri = base_uri .. ('%s/%s'):format(
    uri.encode(foo, uri.PATH_SEGMENT),
    uri.encode(bar, uri.PATH_SEGMENT))
```

But path segments are often have a known format (which doesn't allow `/` inside) and often are just constants. We can write the same code shorter in the case:

```lua
local uri = base_uri .. uri.encode(('%s/%s'):format(foo, bar), uri.PATH)
```

The latter code will be broken by `/` in `foo` or `bar`, but it is fine for cases, when, say, `foo` is an API version and `bar` is API endpoint name -- trusted data with known format:

```lua
local api_version = 'v1'
local api_endpoint = 'send'
<...>
local uri = base_uri .. uri.encode(
    ('%s/%s'):format(api_version, api_endpoint),
    uri.PATH)
```

Or, say:

```lua
local api_version = 'v1'
<...>
local uri = base_uri .. uri.encode(
    ('%s/send'):format(api_version),
    uri.PATH)
```

In a light of the motivation written above I can't understand why `/` is in the unsafe set for `PATH`. It should be in the unsafe set for `PATH_SEGMENT`, not here.

BTW, such motivating examples would be good to include into code comment/docs. It is not always in the active memory and seems helpful. Despite that we discussed it a month or two ago, it was a bit hard for me to recall those details.

Reply:
There is nothing meaningful in escaping `/` in the `uri.PATH` mode, if it is supposed to accept a path (not a path segment). It makes this mode unusable: glance over the examples above. The RFC has neither correct character class definitions, nor an idea how to correctly construct them. So I can't agree: I need and must remember a lot of things to act as a reviewer here. I approved the RFC with words 'the API looks okay, but we need to work on the character classes'. I remember it. And you?